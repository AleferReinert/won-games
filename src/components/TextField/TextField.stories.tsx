import { expect } from '@storybook/jest'
import type { Meta, StoryObj } from '@storybook/react'
import { userEvent, within } from '@storybook/testing-library'
import { Email, Person } from '@styled-icons/material-outlined'
import theme from 'styles/theme'
import { hexToRGBA } from 'utils/tests/helpers'
import TextField from './TextField'

const meta: Meta<typeof TextField> = {
  title: 'Components/Atoms/TextField',
  component: TextField,
  args: {
    name: 'email',
    id: 'email'
  },
  argTypes: {
    onInput: {
      action: 'changed',
      table: { disable: true }
    },
    icon: {
      options: ['None', 'Email', 'Person'],
      mapping: {
        None: '',
        Email: <Email />,
        Person: <Person />
      }
    },
    $iconPosition: { if: { arg: 'icon' } },
    name: { table: { disable: true } },
    id: { table: { disable: true } }
  },
  parameters: {
    backgrounds: {
      default: 'Light'
    }
  }
}
export default meta

type Story = StoryObj<typeof TextField>

export const Default: Story = {
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement)
    const textField = canvas.getByRole('textbox')
    const labels = canvasElement.getElementsByTagName('label')

    expect(textField).toHaveAttribute('id')
    expect(textField).toHaveAttribute('name')

    // without label as default
    expect(labels.length).toBe(0)
  }
}

export const Focus: Story = {
  play: async ({ canvasElement, args }) => {
    const canvas = within(canvasElement)
    const textField = canvas.getByRole('textbox')

    // Accessibility with tab
    expect(document.body).toHaveFocus()
    await userEvent.tab()
    expect(textField).toHaveFocus()

    // change text when typing
    const newText = 'my new custom value'
    await userEvent.type(textField, newText)
    expect(textField).toHaveValue(newText)
    expect(args.onInput).toHaveBeenCalledTimes(newText.length)
    expect(args.onInput).toHaveBeenCalledWith(newText)
  }
}

export const WithLabel: Story = {
  args: {
    label: 'With label'
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement)
    const label = canvas.getByText(/with label/i)

    expect(label).toBeInTheDocument()

    // Input with focus when label is clicked
    await userEvent.click(label)
    const textField = canvas.getByRole('textbox')
    expect(textField).toHaveFocus()

    // reset
    await userEvent.click(document.body)
  }
}

export const WithIcon: Story = {
  args: {
    icon: <Person />
  },
  play: async ({ canvasElement }) => {
    const icon = canvasElement.getElementsByTagName('svg')
    const inputWrapper = icon[0].parentElement?.parentElement

    expect(icon.length).toBe(1)

    // left position as default
    expect(inputWrapper).not.toHaveStyle({ 'flex-direction': 'row-reverse' })
  }
}

export const IconRight: Story = {
  args: {
    icon: 'Person',
    $iconPosition: 'right'
  },
  play: async ({ canvasElement }) => {
    const inputWrapper =
      canvasElement.getElementsByTagName('svg')[0].parentElement?.parentElement

    // left position as default
    expect(inputWrapper).toHaveStyle({ 'flex-direction': 'row-reverse' })
  }
}

export const Disabled: Story = {
  args: {
    disabled: true,
    label: 'Disabled'
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement)
    const label = canvas.getByText(/disabled/i)
    const textField = canvas.getByRole('textbox')

    expect(label).toHaveStyle({
      color: theme.colors.gray,
      cursor: 'default'
    })
    expect(textField).toBeDisabled()
    expect(textField).toHaveStyle({ cursor: 'default' })
    expect(textField.parentElement).toHaveStyle({ cursor: 'default' })

    // not be acessible with tab
    expect(document.body).toHaveFocus()
    await userEvent.tab()
    expect(textField).not.toHaveFocus()

    // does not change when typing
    const newText = 'my new text'
    await userEvent.type(textField, newText)
    expect(textField).not.toHaveValue(newText)
  }
}

export const Error: Story = {
  args: {
    errorMessage: 'Ops...something is wrong'
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement)
    const textField = canvas.getByRole('textbox')
    const inputWrapper = textField.parentElement
    const errorMessage = canvas.getByText(/ops...something is wrong/i)

    expect(errorMessage).toHaveStyle({ color: theme.colors.error })

    expect(inputWrapper).toHaveStyle({
      borderColor: hexToRGBA(theme.colors.error)
    })
  }
}

export const InitialValue: Story = {
  args: {
    initialValue: 'Initial value'
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement)
    const textField = canvas.getByRole('textbox')

    expect(textField).toHaveValue('Initial value')
  }
}
