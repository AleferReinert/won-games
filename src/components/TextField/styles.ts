import styled, { css, DefaultTheme } from 'styled-components'
import { TextFieldProps } from '.'

const wrapperModifiers = {
  iconPositionRight: () => css`
    flex-direction: row-reverse;
  `,
  disabled: (theme: DefaultTheme) => css`
    ${Input}, ${InputWrapper} {
      cursor: not-allowed;
      color: ${theme.colors.gray};

      &::placeholder {
        color: currentColor;
      }
    }
  `,
  error: (theme: DefaultTheme) => css`
    ${InputWrapper} {
      border-color: ${theme.colors.error};
    }

    ${Icon},${Label} {
      color: ${theme.colors.error};
    }
  `
}

export const InputWrapper = styled.div<Pick<TextFieldProps, 'iconPosition'>>`
  ${({ theme, iconPosition }) => css`
    display: flex;
    background: ${theme.colors.lightGray};
    border-radius: 0.2rem;
    padding: 0 ${theme.spacings.xxsmall};
    border: 0.2rem solid;
    border-color: ${theme.colors.lightGray};

    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }

    ${iconPosition === 'right' && wrapperModifiers.iconPositionRight()}
  `}
`

export const Input = styled.input`
  ${({ theme }) => css`
    color: ${theme.colors.black};
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall};
    background: transparent;
    border: 0;
    outline: none;
    width: 100%;
  `}
`

export const Label = styled.label`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.small};
    color: ${theme.colors.black};
    cursor: pointer;
  `}
`

export const Icon = styled.span`
  ${({ theme }) => css`
    display: flex;
    width: 2rem;
    color: ${theme.colors.gray};
    margin: 0 ${theme.spacings.xxsmall};

    > svg {
      width: 100%;
    }
  `}
`
export const Error = styled.span`
  ${({ theme }) => css`
    color: ${theme.colors.error};
    font-size: ${theme.font.sizes.xsmall};
  `}
`

type WrapperProps = Pick<TextFieldProps, 'disabled'> & { error?: boolean }

export const Wrapper = styled.div<WrapperProps>`
  ${({ theme, error, disabled }) => css`
    ${!!error && wrapperModifiers.error(theme)}
    ${disabled && wrapperModifiers.disabled(theme)}
  `}
`
