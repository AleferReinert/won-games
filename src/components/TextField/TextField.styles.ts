import { lighten } from 'polished'
import styled, { css, DefaultTheme } from 'styled-components'
import { TextFieldProps } from './TextField'

const modifiers = {
  $iconPositionRight: () => css`
    flex-direction: row-reverse;
  `,
  disabled: (theme: DefaultTheme) => css`
    ${InputWrapper} {
      cursor: default;

      ${Input} {
        cursor: default;
        color: ${lighten(0.15, theme.colors.gray)};

        &::placeholder {
          color: ${lighten(0.15, theme.colors.gray)};
        }
      }

      svg {
        fill: ${lighten(0.15, theme.colors.gray)};
      }
    }

    ${Label} {
      cursor: default;
      color: ${theme.colors.gray};
    }
  `,
  errorMessage: (theme: DefaultTheme) => css`
    border-color: ${theme.colors.error};
  `
}

export const InputWrapper = styled.div<Pick<TextFieldProps, '$iconPosition' | '$errorMessage'>>`
  ${({ theme, $errorMessage, $iconPosition }) => css`
    display: flex;
    align-items: center;
    background: ${theme.colors.lightGray};
    border-radius: 0.2rem;
    padding: 0 ${theme.spacings.xxsmall};
    border: 0.1rem solid ${theme.colors.lightGray};
    margin-bottom: ${theme.spacings.xxsmall};

    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }

    svg {
      fill: ${theme.colors.gray};
      width: 2.4rem;
      height: 2.4rem;
    }

    ${$iconPosition === 'right' && modifiers.$iconPositionRight()}
    ${!!$errorMessage && modifiers.errorMessage(theme)}
  `}
`

export const Input = styled.input`
  ${({ theme }) => css`
    color: ${theme.colors.black};
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall};
    background: transparent;
    border: 0;
    outline: none;
    width: 100%;

    &:-webkit-autofill {
      -webkit-box-shadow: 0 0 0px 1000px ${theme.colors.lightGray} inset;
    }

    &::placeholder {
      color: ${theme.colors.gray};
    }
  `}
`

export const Label = styled.label`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.small};
    color: ${theme.colors.black};
    cursor: pointer;
  `}
`

export const Icon = styled.span`
  ${({ theme }) => css`
    display: flex;
    color: ${theme.colors.gray};
    margin: 0 ${theme.spacings.xxsmall};
  `}
`
export const ErrorMessage = styled.span`
  ${({ theme }) => css`
    color: ${theme.colors.error};
    font-size: ${theme.font.sizes.xsmall};
  `}
`

export const Wrapper = styled.div<Pick<TextFieldProps, 'disabled'>>`
  ${({ theme, disabled }) => css`
    ${disabled && modifiers.disabled(theme)}
  `}
`
