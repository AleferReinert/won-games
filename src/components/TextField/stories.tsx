import { ComponentStory, ComponentMeta } from '@storybook/react'
import TextField from '.'
import { Email } from '@styled-icons/material-outlined'

export default {
  title: 'Forms/TextField',
  component: TextField,
  args: {
    label: 'E-mail',
    labelFor: 'email',
    id: 'email',
    initialValue: '',
    placeholder: 'example@gmail.com',
    icon: <Email />
  },
  argTypes: {
    onInput: { action: 'changed' }
  },
  parameters: {
    backgrounds: {
      default: 'Light'
    }
  }
} as ComponentMeta<typeof TextField>

export const Default: ComponentStory<typeof TextField> = (args) => (
  <TextField {...args} />
)

export const WithError: ComponentStory<typeof TextField> = (args) => (
  <TextField {...args} />
)

WithError.args = {
  error: 'Ops...something is wrong'
}
