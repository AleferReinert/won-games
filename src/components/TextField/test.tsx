import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import TextField from '.'
import { renderWithTheme } from 'utils/tests/helpers'
import { Email } from '@styled-icons/material-outlined'

describe('<TextField />', () => {
  it('should render with label', () => {
    renderWithTheme(<TextField label='Label' labelFor='Field' id='Field' />)
    expect(screen.getByText('Label')).toBeInTheDocument()
  })

  it('should render without label', () => {
    renderWithTheme(<TextField />)
    expect(screen.queryByLabelText('Label')).not.toBeInTheDocument()
  })

  it('should render with placeholder', () => {
    renderWithTheme(<TextField placeholder='lorem ipsum' />)
    expect(screen.queryByPlaceholderText('lorem ipsum')).toBeInTheDocument()
  })

  it('Change its value when typing', async () => {
    const onInput = jest.fn()
    renderWithTheme(
      <TextField
        label='TextField'
        labelFor='TextField'
        onInput={onInput}
        id='TextField'
      />
    )
    const input = screen.getByRole('textbox')
    const text = 'My new text'

    userEvent.type(input, text)
    await waitFor(() => {
      expect(input).toHaveValue(text)
      expect(onInput).toHaveBeenCalledTimes(text.length)
    })
    expect(onInput).toHaveBeenCalledWith(text)
  })

  it('does not change its value when disabled', async () => {
    const onInput = jest.fn()
    renderWithTheme(
      <TextField
        disabled
        onInput={onInput}
        label='TextField'
        labelFor='TextField'
        id='TextField'
      />
    )
    const input = screen.getByRole('textbox')
    const text = 'New text'

    expect(input).toBeDisabled()
    userEvent.type(input, text)
    await waitFor(() => {
      expect(input).not.toHaveValue(text)
    })
    expect(onInput).not.toHaveBeenCalled()
  })

  it('should render with left icon as default', () => {
    renderWithTheme(<TextField icon={<Email data-testid='icon' />} />)

    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })

  it('should render with right icon', () => {
    renderWithTheme(
      <TextField iconPosition='right' icon={<Email data-testid='icon' />} />
    )

    expect(
      screen.getByRole('textbox').parentElement as HTMLElement
    ).toHaveStyle({ flexDirection: 'row-reverse' })
  })

  it('should render disabled', () => {
    renderWithTheme(<TextField disabled />)

    expect(screen.getByRole('textbox')).toHaveAttribute('disabled')
  })

  it('should render with error message', async () => {
    const { container } = renderWithTheme(
      <TextField
        icon={<Email data-testid='icon' />}
        label='TextField'
        labelFor='TextField'
        error='Error message'
      />
    )

    expect(screen.getByText(/error message/i)).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('should be acessible with tab', async () => {
    renderWithTheme(
      <TextField label='TextField' labelFor='TextField' id='TextField' />
    )

    expect(document.body).toHaveFocus()
    await userEvent.tab()
    expect(screen.getByLabelText('TextField')).toHaveFocus()
  })

  it('should not be acessible with tab when disabled', async () => {
    renderWithTheme(
      <TextField
        disabled
        label='TextField'
        labelFor='TextField'
        id='TextField'
      />
    )
    const input = screen.getByRole('textbox')
    expect(document.body).toHaveFocus()

    userEvent.tab()
    expect(input).not.toHaveFocus()
  })
})
