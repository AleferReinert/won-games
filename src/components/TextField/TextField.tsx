import { ComponentProps, ReactNode } from 'react'
import * as S from './TextField.styles'

export interface TextFieldProps extends ComponentProps<'input'> {
  name: string
  label?: string
  icon?: ReactNode
  $iconPosition?: 'left' | 'right'
  disabled?: boolean
  $errorMessage?: string
  initialValue?: string
}

const TextField = ({
  name,
  label,
  icon,
  $iconPosition = 'left',
  disabled = false,
  $errorMessage,
  // initialValue = '',
  ...props
}: TextFieldProps) => {
  // const [value, setValue] = useState(initialValue)

  // const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
  //   const newValue = e.currentTarget.value
  //   setValue(newValue)
  // }

  return (
    <S.Wrapper disabled={disabled}>
      {!!label && <S.Label htmlFor={name}>{label}</S.Label>}
      <S.InputWrapper $iconPosition={$iconPosition} $errorMessage={$errorMessage}>
        {!!icon && <S.Icon>{icon}</S.Icon>}
        <S.Input
          type='text'
          // onChange={onChange}
          // value={value}
          disabled={disabled}
          autoComplete='off'
          name={name}
          id={name}
          {...props}
        />
      </S.InputWrapper>
      {!!$errorMessage && <S.ErrorMessage>{$errorMessage}</S.ErrorMessage>}
    </S.Wrapper>
  )
}

export default TextField
