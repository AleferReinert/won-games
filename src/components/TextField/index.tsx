import * as S from './styles'
import { useState, InputHTMLAttributes, ReactNode } from 'react'

export type TextFieldProps = {
  onInput?: (value: string) => void
  label?: string
  labelFor?: string
  initialValue?: string
  icon?: ReactNode
  iconPosition?: 'left' | 'right'
  disabled?: boolean
  error?: string
} & InputHTMLAttributes<HTMLInputElement>

const TextField = ({
  onInput,
  label,
  labelFor = '',
  initialValue = '',
  icon,
  iconPosition = 'left',
  disabled = false,
  error,
  ...props
}: TextFieldProps) => {
  const [value, setValue] = useState(initialValue)

  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const newValue = e.currentTarget.value
    setValue(newValue)

    !!onInput && onInput(newValue)
  }

  return (
    <S.Wrapper error={!!error} disabled={disabled}>
      {!!label && <S.Label htmlFor={labelFor}>{label}</S.Label>}
      <S.InputWrapper iconPosition={iconPosition}>
        {!!icon && <S.Icon>{icon}</S.Icon>}
        <S.Input
          type='text'
          onChange={onChange}
          value={value}
          disabled={disabled}
          autoComplete='off'
          {...props}
        />
      </S.InputWrapper>
      {!!error && <S.Error>{error}</S.Error>}
    </S.Wrapper>
  )
}

export default TextField
