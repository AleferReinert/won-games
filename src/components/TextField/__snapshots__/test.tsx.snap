// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> should render with error message 1`] = `
.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: #EAEAEA;
  border-radius: 0.2rem;
  padding: 0 0.8rem;
  border: 0.2rem solid;
  border-color: #EAEAEA;
}

.c4:focus-within {
  box-shadow: 0 0 0.5rem #F231A5;
}

.c8 {
  color: #030517;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  padding: 0.8rem;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

.c8:-webkit-autofill {
  -webkit-box-shadow: 0 0 0px 1000px #EAEAEA inset;
}

.c2 {
  font-size: 1.4rem;
  color: #030517;
  cursor: pointer;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 2rem;
  color: #8F8F8F;
  margin: 0 0.8rem;
}

.c6 > svg {
  width: 100%;
}

.c9 {
  color: #ff6347;
  font-size: 1.2rem;
}

.c0 .c3 {
  border-color: #ff6347;
}

.c0 .c5,
.c0 .c1 {
  color: #ff6347;
}

.c7 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

<div>
  <div
    class="c0"
  >
    <label
      class="c1 c2"
      for="TextField"
    >
      TextField
    </label>
    <div
      class="c3 c4"
    >
      <span
        class="c5 c6"
      >
        <svg
          aria-hidden="true"
          class="c7"
          data-testid="icon"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M22 6c0-1.1-.9-2-2-2H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6zm-2 0-8 5-8-5h16zm0 12H4V8l8 5 8-5v10z"
          />
        </svg>
      </span>
      <input
        autocomplete="off"
        class="c8"
        type="text"
        value=""
      />
    </div>
    <span
      class="c9"
    >
      Error message
    </span>
  </div>
</div>
`;
